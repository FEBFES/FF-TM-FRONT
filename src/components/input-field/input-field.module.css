.field {
    display: flex;
    flex-direction: column;
    width: 100%;
    height: 90px;
}

.inputLabel {
    font-size: 12px;
    font-weight: 300;
    margin-bottom: 4px;
}

.input {
    outline: none;
    background: #f6f6f6;
    color: black;
    font-size: 14px;
    font-weight: 100;
    padding: 12px 20px;
    width: 100%;
    height: 48px;
    border: 1px solid rgb(223, 223, 223);
    border-radius: 8px;
    -webkit-border-radius: 8px;
    -moz-border-radius: 8px;
    -ms-border-radius: 8px;
    -o-border-radius: 8px;
    transition: .4s;
    -webkit-transition: .4s;
    -moz-transition: .4s;
    -ms-transition: .4s;
    -o-transition: .4s;  
}

.input_invalid {
    border: 1px solid rgb(162, 8, 8);
    box-shadow: 0px 0px 0px 4px rgba(219, 0, 0, 0.1);
}
.input_valid:hover {
    border: 1px solid rgba(0, 58, 219, 0.5);
    box-shadow: 0px 0px 0px 4px rgba(0, 58, 219, 0.1);
}
.input_valid:focus {
    border: 1px solid rgba(0, 58, 219, 0.5);
}

.errorLabel {
    margin-top: 4px;
    font-size: 12px;
    color: rgba(219, 0, 0, 0.659);
}