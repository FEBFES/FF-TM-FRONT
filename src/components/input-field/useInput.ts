import { useEffect, useState } from 'react';

export type ValidationRule = {
  type: 'required' | 'minLength' | 'maxLength' | 'pattern';
  value?: number | RegExp;
  errorMsg: string;
};

type ValidationRules = ValidationRule[]

const useInput = (currentInputValue: string, validationRules: ValidationRules) => {
  const [error, setError] = useState<string | null>(null)
    const [isValid, setIsValid] = useState<boolean>(true)

    useEffect(() => {
        if (currentInputValue) validate(currentInputValue)
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [currentInputValue])

  const validate = (value: string): boolean => {
    for (const rule of validationRules) {
      let isValid = true;

      switch (rule.type) {
        case 'required':
          isValid = !!value.trim();
          break;
        case 'minLength':
          isValid = value.length >= (rule.value as number);
          break;
        case 'maxLength':
          isValid = value.length <= (rule.value as number);
          break;
        case 'pattern':
          isValid = (rule.value as RegExp).test(value);
          break;
        default:
          break;
    }

      setIsValid(isValid)
      if (!isValid) {
        setError(rule.errorMsg)
        return false
      }
    }

    setError(null);
    return true;
  };

  const handleBlur = () => {
    validate(currentInputValue);
  };

//   const reset = () => {
//     setValue(currentInputValue);
//     setError(null);
//   };

  return {
    onBlur: handleBlur,
    error,
    isValid
  };
};

export default useInput;