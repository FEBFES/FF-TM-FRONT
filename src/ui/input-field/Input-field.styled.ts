import styled from 'styled-components';

export const SInputContainer = styled.div`
  margin-top: 10px;
  position: relative;
  height: 46px;
  width: 100%;
`;

export const SInput = styled.input`
  outline: none;
  background: transparent;
  border: 1px solid var(--bg-border);
  width: 100%;
  height: 100%;
  border-radius: 8px;
  padding: 16px 10px 0 10px;
  font-size: 14px;
  color: var(--font-defautl);
`;

export const SInputLabel = styled.label<{ isActive: boolean }>`
  font-size: ${({ isActive }) => (isActive ? '12px' : '14px')};
  transition: 0.1s;
  position: absolute;
  left: ${({ isActive }) => (isActive ? '12px' : '10px')};
  top: ${({ isActive }) => (isActive ? '16px' : '50%')};
  transform: translateY(-50%);
  color: var(--font-gray);
  background: ${({ isActive }) => isActive && 'transparent'};
`;

export const SPasswordIcon = styled.div`
  cursor: pointer;
  position: absolute;
  right: 10px;
  top: 50%;
  transform: translateY(-50%);
`;
