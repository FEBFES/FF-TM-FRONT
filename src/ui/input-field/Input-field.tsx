import React, {
  DetailedHTMLProps,
  InputHTMLAttributes,
  useEffect,
  useState,
} from 'react';
import { faEye } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  SPasswordIcon,
  SInputContainer,
  SInput,
  SInputLabel,
} from './Input-field.styled';

interface IInputFieldProps
  extends DetailedHTMLProps<
    InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  > {
  value?: any;
  withLabel?: boolean;
  containerStyle?: string;
}

export const InputField: React.FC<IInputFieldProps> = ({
  type,
  value,
  withLabel = true,
  className,
  ...props
}): JSX.Element => {
  const [showPassword, setShowPassword] = useState(false);
  const [inputActive, setInputActive] = useState<boolean>(!!value);

  useEffect(() => {
    if (withLabel) {
      setInputActive(!!value);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value]);

  return (
    <SInputContainer
      onClick={() => {
        if (withLabel) setInputActive(true);
      }}
      onBlur={() => {
        if (!value && withLabel) {
          setInputActive(false);
        }
      }}
    >
      <SInput
        value={value}
        type={showPassword ? 'text' : type}
        className={className}
        {...props}
        placeholder={withLabel ? '' : props.placeholder}
        id={'input'}
      />
      {withLabel && (
        <SInputLabel isActive={inputActive} htmlFor="input">
          {props.placeholder}
        </SInputLabel>
      )}
      {type === 'password' && value !== '' && (
        <SPasswordIcon
          onMouseDown={() => setShowPassword(true)}
          onMouseUp={() => setShowPassword(false)}
        >
          <FontAwesomeIcon icon={faEye} />
        </SPasswordIcon>
      )}
    </SInputContainer>
  );
};
