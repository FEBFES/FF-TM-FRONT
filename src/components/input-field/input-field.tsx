import { ChangeEventHandler, InputHTMLAttributes } from "react"
import styles from "./input-field.module.css"
import useInput, { ValidationRule } from "./useInput"
import { clsx } from "clsx"

interface InputFieldProps extends InputHTMLAttributes<HTMLInputElement> {
	label?: string
	type?: string
	placeholder?: string
	validationRules?: ValidationRule[]
	value?: string
	onChange?: ChangeEventHandler<HTMLInputElement>
	onBlur?: () => void
}

const InputField: React.FC<InputFieldProps> = ({
	label,
	type = "text",
	placeholder,
	validationRules,
	value,
	onChange,
	onBlur,
	...props
}): JSX.Element => {
	const {
		onBlur: onBlurValidate,
		error: validateError,
		isValid,
	} = useInput(value || "", validationRules || [])

	return (
		<div className={styles.field}>
			{label && (
				<label className={styles.inputLabel} htmlFor={props.id || "input"}>
					{label}
				</label>
			)}

			<input
				{...props}
				id={props.id || "input"}
				type={type}
				value={value}
				onChange={onChange}
				onBlur={() => {
					if (onBlur) onBlur()
					onBlurValidate()
				}}
				placeholder={placeholder}
				className={clsx(
					styles.input,
					isValid && styles.input_valid,
					!isValid && styles.input_invalid
				)}
			/>

			{!isValid && <span className={styles.errorLabel}>{validateError}</span>}
		</div>
	)
}
InputField.displayName = "InputField"

export default InputField
