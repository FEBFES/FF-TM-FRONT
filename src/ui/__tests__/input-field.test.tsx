import React from 'react';
import { InputField } from '../input-field/Input-field';
import renderer from 'react-test-renderer';

it('SNAPSHOTS: UI - InputField component - render', () => {
  const elem = renderer.create(<InputField value={'asdas'} />).toJSON();
  expect(elem).toMatchSnapshot();
});
