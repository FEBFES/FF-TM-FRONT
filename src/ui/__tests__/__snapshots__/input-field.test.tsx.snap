// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOTS: UI - InputField component - render 1`] = `
<div
  className="sc-bcPKhP kvwdkc"
  onBlur={[Function]}
  onClick={[Function]}
>
  <input
    className="sc-grXZZQ dWqGUc"
    id="input"
    placeholder=""
    value="asdas"
  />
  <label
    className="sc-dkjKgF dwxDqy"
    htmlFor="input"
    isActive={true}
  />
</div>
`;
